### PR TITLE
[dart-dio-next] Improve handling nullable objects

### PR DESCRIPTION
**Pull request by ahmednfwela**
Currently nullable types need to be handled better at serialization level, for example:
deserializing `BuiltMap<String,JsonObject>` when one of the values is `null` will throw an error at run time, since `JsonObject` is not a nullable type

these changes are not perfect, especially changing these
```java
typeMapping.put("object", "JsonObject?");
typeMapping.put("AnyType", "JsonObject?");

imports.put("JsonObject?", "package:built_value/json_object.dart");
```
but I am not sure what's the best approach for putting the null operator (`?`) after a type ?

also this PR makes use of the `FullType.nullable` named constructor for defining serializers

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh bin/configs/dart*
  ./bin/utils/export_docs_generators.sh
  ```  
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.3.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

CC: @jaumard @josh-burton @amondnet @sbu-WBT @kuhnroyal @agilob

### CODE REVIEW

**Comment by kuhnroyal** [comment]
Not everything is nullable, we need to default to non-nullable.
This requires built_value >= 8.1.0

Do you have a sample spec for collections with nullable content?

**Comment by kuhnroyal** [comment]
I created https://github.com/OpenAPITools/openapi-generator/pull/10122 for built_value update

**Comment by kuhnroyal** [comment]
Looking at Kotlin and C# implementations, none of them support null in collections. There also doesn't seem to be a representation for this in the specification.

Maybe a ref with `nullable: true` not sure.
I don't think we should merge this atm.

**Comment by ahmednfwela** [comment]
@kuhnroyal but there is a spec for this https://github.com/OpenAPITools/openapi-generator/blob/cd4e7adb127e9af7fb0c2be2bdcb1d623d5a15b1/samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/lib/src/model/nullable_class.dart#L196

**Comment by ahmednfwela** [comment]
The only problem I am facing is:
how to use `JsonObject` for non-nullable objects, and `JsonObject?` for nullable objects

**Comment by kuhnroyal** [comment]
I see, I didn't know that `items` can have `nullable` property.
You can try accessing this via `CodegenProperty.items.isNullable` or `CodegenProperty.mostInnerItems.isNullable` in the template. I don't think you need to change anything in the Java code. We definitely don't want `JsonObject?` in the type or import mapping.

**Comment by ahmednfwela** [comment]
@kuhnroyal what do you think about this change? I even removed the java code for the serializer type and wrote it in mustache using recursive patterns

**Comment by ahmednfwela** [comment]
also @wing328 can this be included in 5.2.1 ?

**Comment by kuhnroyal** [comment]
I like it but there is still a couple things open. I don't think this should go into 5.2.1

**Comment by ahmednfwela** [comment]
@kuhnroyal is there anything else I can fix on my end?

File: samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/README.md [hunk 1]
```diff
@@ -64,97 +64,97 @@ All URIs are relative to *http://petstore.swagger.io:80/v2*
...
-[*UserApi*](doc/UserApi.md) | [**createUsersWithArrayInput**](doc/UserApi.md#createuserswitharrayinput) | **POST** /user/createWithArray | Creates list of users with given input array
-[*UserApi*](doc/UserApi.md) | [**createUsersWithListInput**](doc/UserApi.md#createuserswithlistinput) | **POST** /user/createWithList | Creates list of users with given input array
-[*UserApi*](doc/UserApi.md) | [**deleteUser**](doc/UserApi.md#deleteuser) | **DELETE** /user/{username} | Delete user
-[*UserApi*](doc/UserApi.md) | [**getUserByName**](doc/UserApi.md#getuserbyname) | **GET** /user/{username} | Get user by user name
-[*UserApi*](doc/UserApi.md) | [**loginUser**](doc/UserApi.md#loginuser) | **GET** /user/login | Logs user into the system
-[*UserApi*](doc/UserApi.md) | [**logoutUser**](doc/UserApi.md#logoutuser) | **GET** /user/logout | Logs out current logged in user session
-[*UserApi*](doc/UserApi.md) | [**updateUser**](doc/UserApi.md#updateuser) | **PUT** /user/{username} | Updated user
+[*AnotherFakeApi*](doc\AnotherFakeApi.md) | [**call123testSpecialTags**](doc\AnotherFakeApi.md#call123testspecialtags) | **PATCH** /another-fake/dummy | To test special tags
```
Comment (agilob) [inline samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/README.md:67]: this needs reverting

File: samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/lib/src/model/animal.dart [hunk 2]
```diff
@@ -13,10 +13,10 @@ part 'animal.g.dart';
 /// * [className] 
 /// * [color] 
 abstract class Animal implements Built<Animal, AnimalBuilder> {
-    @BuiltValueField(wireName: r'className')
+    @BuiltValueField(wireName: r'className')
```
Comment (agilob) [inline samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/lib/src/model/animal.dart:16]: remove extra whitespaces at the end

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartDioNextClientCodegen.java [hunk 3]
```diff
@@ -276,35 +276,43 @@ public void postProcessModelProperty(CodegenModel model, CodegenProperty propert
...
-            sb.append("String), FullType(");
-        }
-        if (property.items.isContainer) {
-            appendBuiltValueCollection(sb, property.items);
-        } else {
-            sb.append(property.items.datatypeWithEnum);
-        }
-        sb.append(")]");
-    }
+    // private String createBuiltValueSerializerType(CodegenProperty property) {
```
Comment (agilob) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartDioNextClientCodegen.java:283]: dont commit commented out code, delete it instead

File: samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/lib/src/model/additional_properties_class.dart [hunk 4]
```diff
@@ -46,13 +46,13 @@ class _$AdditionalPropertiesClassSerializer implements StructuredSerializer<Addi
             result
                 ..add(r'map_property')
                 ..add(serializers.serialize(object.mapProperty,
-                    specifiedType: const FullType(BuiltMap, [FullType(String), FullType(String)])));
+                    specifiedType: const FullType(BuiltMap, [const FullType(String), const FullType(String)]) ,),);
```
Comment (kuhnroyal) [inline samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/lib/src/model/additional_properties_class.dart:49]: Can you try to remove the additional `const` keywords which may generate analyzer warnings for some users?
And removing the additional whitespaces and commas would also be great for the diff.

File: samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/lib/src/model/additional_properties_class.dart [hunk 5]
```diff
@@ -15,10 +15,10 @@ part 'additional_properties_class.g.dart';
 /// * [mapOfMapProperty] 
 abstract class AdditionalPropertiesClass implements Built<AdditionalPropertiesClass, AdditionalPropertiesClassBuilder> {
     @BuiltValueField(wireName: r'map_property')
-    BuiltMap<String, String>? get mapProperty;
```
Comment (kuhnroyal) [inline samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/lib/src/model/additional_properties_class.dart:18]: Add the space back in

**Review by kuhnroyal — CHANGES_REQUESTED** [review]
Really like what you did here! 2 small changes requested

File: samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/lib/src/model/additional_properties_class.dart (hunk 4 above)
Comment (ahmednfwela) [reply samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/lib/src/model/additional_properties_class.dart:49]: for the commas, it's good if users (like me) are formatting the generated output

Comment (kuhnroyal) [reply samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/lib/src/model/additional_properties_class.dart]: I agree but can we do that in a separate PR?

Comment (ahmednfwela) [reply samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/lib/src/model/additional_properties_class.dart]: sure, why not

File: modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/class.mustache [hunk 6]
```diff
@@ -92,21 +92,24 @@ class _${{classname}}Serializer implements StructuredSerializer<{{classname}}> {
             final key = iterator.current as String;
             iterator.moveNext();
             final Object? value = iterator.current;
+            
             switch (key) {
                 {{#vars}}
-                case r'{{baseName}}':
+                case r'{{baseName}}':
```
Comment (kuhnroyal) [inline modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/class.mustache:98]:
```suggestion
                case r'{{baseName}}':
```

File: modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/class.mustache [hunk 7]
```diff
@@ -67,15 +67,15 @@ class _${{classname}}Serializer implements StructuredSerializer<{{classname}}> {
...
+                specifiedType: const {{>serialization/built_value/variable_serializer_type}}));
         {{/required}}
         {{^required}}
         if (object.{{{name}}} != null) {
             {{! Non-required properties are only serialized if not null. }}
             result
                 ..add(r'{{baseName}}')
                 ..add(serializers.serialize(object.{{{name}}},
-                    specifiedType: {{{vendorExtensions.x-built-value-serializer-type}}}));
+                    specifiedType: const {{>serialization/built_value/variable_serializer_type}}));
```
Comment (kuhnroyal) [inline modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/class.mustache:78]:
```suggestion
                    specifiedType: const {{>serialization/built_value/variable_serializer_type}}));
```

**Review by kuhnroyal — APPROVED** [review]
Awesome work, that will also make it easier to fix a couple problems around missing serializers. Thanks!

**Comment by kuhnroyal** [comment]
@agilob Do you have anything else here?

**Review by agilob — APPROVED** [review]
LGTM

**Comment by kuhnroyal** [comment]
@wing328 This can be merged

**Pull request merged**